src/server/tran_server.cpp

```diff
@@ -484,6 +488,10 @@ tran_server::connection_handler::disconnect_async (bool with_disc_msg)
   assert (m_state == state::CONNECTING || m_state == state::CONNECTED);
   m_state = state::DISCONNECTING;
 
+  ulock.unlock();
+
+  on_disconnecting (); // server-type specific jobs before disconnect
```

Comment:
Honest question.
Would it be better to also call this in the async functor below?
This would lead to a more unified approach. The function `disconnect_async` just sets the state to `DISCONNECTING` synchronously, while deferring everything to the async functor.
Later on, it might even be needed to alter the order where the `on_disconnecting` is called within the disconnect functor as a whole.

Moreover, as we have these new functions now, their semantics is more like:
- `after_connecting`
- `before_disconnecting`

That's how it looks to me. Is it worthwhile to alter their name?

Reply:
> The function disconnect_async just sets the state to DISCONNECTING synchronously, while deferring everything to the async functor.

Hmm, that's what I did, to begin with. now, it doesn't matter so I moved here since it just looks more natural. However, I'll move it to the async functor again, because it has less constraint in the location.

> after_connecting
before_disconnecting

I regard what's done are on_connecting and on_disconnecting as the connecting/disconnecting jobs. Without it, there is nothing but transition in that state. What I'm thinking of for the states in detail:
- IDLE -> CONNECTING: establish the internal connection.
- on CONNECTING: preliminary jobs before opening the connection to the outside such as the catch-up, adding prior_sink_hooks
- on DISCONNECTING: preliminary jobs before closing the connection such as removing sink hooks.
- DISCONNECTING -> IDLE: destroy the internal connection. 

Doesn't it make sense?

Reply:
Yes, it makes sense.
That's how the structure is now in `tran_server::connection_handler::connect` and `tran_server::connection_handler::disconnect_async` and based on this, the `after/before_disconnecting` makes more sense to me.

Maybe I have a bias from Delphi where there is something like - https://docwiki.embarcadero.com/Libraries/Sydney/en/System.TObject_Methods :)
But, [what's in a name](https://writingtips.org/whats-in-a-name/) ?